app/models/accounting/quickbooks/full_fetcher.rb

```diff
@@ -10,20 +10,56 @@ def initialize(qb_connection = Division.root.qb_connection)
       end
 
       def fetch_all
-        divisions_accounts = Division.clear_all_accounts
+        divisions_accounts = clear_all_accounts!
 
         ::Accounting::Transaction.destroy_all
         ::Accounting::Account.destroy_all
 
         started_fetch_at = Time.zone.now
 
-        ::Accounting::Quickbooks::AccountFetcher.new.fetch
-        ::Accounting::Quickbooks::TransactionFetcher.new.fetch
+        ::Accounting::Quickbooks::AccountFetcher.new(@qb_connection).fetch
+        ::Accounting::Quickbooks::TransactionFetcher.new(@qb_connection).fetch
 
         qb_connection.update_attribute(:last_updated_at, started_fetch_at)
 
-        Division.restore_all_accounts(divisions_accounts)
+        restore_all_accounts!(divisions_accounts)
       end
+
+      private
+
+      # Set all divisions' accounts to nil and return a hash of the QB ids of the removed accounts
+      # by division
+      def clear_all_accounts!
+        divisions_accounts = {}
+        Division.all.each do |d|
+          accounts_qb_ids = {
+            principal_qb_id: d.principal_account&.qb_id,
+            interest_receivable_qb_id: d.interest_receivable_account&.qb_id,
+            interest_income_qb_id: d.interest_income_account&.qb_id,
+          }
+          d.update(
+            principal_account_id: nil,
+            interest_receivable_account_id: nil,
+            interest_income_account_id: nil,
+          )
+          divisions_accounts[d.id] = accounts_qb_ids
+        end
+        divisions_accounts
+      end
+
+      # Restore all divisions' accounts to the ones passed in. Argument is expected to be a hash of
+      # the form returned by `#clear_all_accounts!`. NOTE: If a previously selected account no
+      # longer exists after the full sync from QB, it will be set to nil. This is by design.
+      def restore_all_accounts!(divisions_accounts)
+        divisions_accounts.each do |did, qb_ids|
+          Division.find(did).update(
+            principal_account: Accounting::Account.find_by(qb_id: qb_ids[:principal_qb_id]),
+            interest_receivable_account: Accounting::Account.find_by(qb_id: qb_ids[:interest_receivable_qb_id]),
+            interest_income_account: Accounting::Account.find_by(qb_id: qb_ids[:interest_income_qb_id]),
+          )
+        end
+      end
+
```

Comment:
Shouldn't we be deleting transactions here too? I realize the spec didn't say one way or the other but I think transactions and accounts should both be deleted. @linesarefuzzy can you confirm this?

Reply:
They all get deleted in the fetch_all method above. (I think I accidentally committed that to the release branch so it’s not shown as a change here.)